### ISSUE TITLE
Read resources refresh state of terraform

### ISSUE THREAD

**Issue by vbauzysvmware** [terraform-providers/terraform-provider-vcd#239]
Ref: https://github.com/terraform-providers/terraform-provider-vcd/pull/236
Investigate pros and cons to on reading function in resource to update state with current state in vCD by overriding values. Or keep them. And/Or adding computed/additional values which are provided additionally.

Currently major of implemented resources only check if resource exist and doesn't update state of changed values.

**Comment by lvirbalas** [comment]
One of the priorities in this task is to add a read option for the following field:
- `vcd_org_vdc.compute_capacity.used` to reflect how much of the VDC allocation is used currently.

**Comment by buec011** [comment]
Hello,
I'm curious, is this bug for the new vcd_org_vdc ressourceonly? 
Or does it apply for all resources that can be build with terraform. 
From title "Read resources refresh state of terraform" and description "Currently major of implemented resources only check if resource exist and doesn't update state of changed values."
it applies to more ressource e.g. maybe vcd_vapp_vm, vcd_vapp_network, vcd_network ...

For us it is very important that *all* configurations that can be deployed with terraform will be checked and refreshed. Otherwise terraform plan/apply may have unexpected results and would render the main idea of terraform useless...

**Comment by lvirbalas** [comment]
Hello,

> I'm curious, is this bug for the new vcd_org_vdc ressourceonly?

We logged this issue as a parent one for capturing implementations of adding “read” capabilities, and even more generally to have it in the backlog.

> it applies to more ressource e.g. maybe vcd_vapp_vm, vcd_vapp_network, vcd_network ...

Please note, that we are developing this provider actively and many new resources & features are being added with each release. In addition, to satisfy the most actual needs of users we are releasing _often_ as you can see here (look from v2.0.0 and up when we started developing it): https://github.com/terraform-providers/terraform-provider-vcd/releases

In effect, this means that the most important and requested aspects are implement (or merged from contributors) and released first, after which "deeper" Terraform capabilities are being added. So far the users have been happy about this quick-cycle approach.

> For us it is very important that all configurations that can be deployed with terraform will be checked and refreshed.

Please explain your use case in more detail. Why exactly this is so important? Are you changing parts of deployed configuration "behind the back" of Terraform? If yes, which resources and which parameters and why?

We started adding read capabilities incrementally to the most important places, like networking (IP addresses). In the near future we are planning to add many data sources and during that effort the plan is to cover reads of most of the parameters. But to be fair, reads only add *visibility* if someone changed some configuration _not_ through Terraform, but that's about it.

**Comment by buec011** [comment]
> Please explain your use case in more detail. Why exactly this is so important? Are you changing parts of deployed configuration "behind the back" of Terraform? If yes, which resources and which parameters and why?
> 

One example is a shared environment with a customer. Even if we told him we have a Terraform integration, someone changed configuration "behind the back" of Terraform.
I can imagine cases where our own standby does the "quick way" via cloud methods to solve an issue, and our Terraform Core team needs to adjust the config at next business day.

We manage many clouds and we set up a nightly-check (more or less a terraform plan) to detect and report changes that have been done "behind the back" of Terraform

This is the reason I cannot tell you a special resource. So I would like to suggest to add a "read" for every option you can set via terraform. Maybe if you add a new item we can configure via terraform, add the "read" immediate with it?

Anyway I think this have a synergetic effect with a import function for ressources? 
(https://github.com/terraform-providers/terraform-provider-vcd/issues/252)
To refresh the state of a setting, or to import a resource, you need the "read" part implemented? Or am I wrong?

**Comment by lvirbalas** [comment]
Thank you for explaining - it is much clearer now.

> So I would like to suggest to add a "read" for every option you can set via terraform.

Yes, we're actually doing that, but it takes time to build it out iteratively... Hence, to sum it up, for your use case of monitoring back-end changes through Terraform, the later 2.5.0 release will be needed. In that release we are planning to dive wide & deep into (a) datasources, (b) import capabilities and with synergy to (c) the reads as well.

**Comment by buec011** [comment]
> In that release we are planning to dive wide & deep into (a) datasources, (b) import capabilities and with synergy to (c) the reads as well.

It is nice to hear that and all I wanted to make sure :-) .

**Comment by lvirbalas** [comment]
Please let us know if you'd be interested in trying those additions early (before official release or even while PR is still open and not merged). This is an open-source project after all :)

**Comment by vbauzysvmware** [comment]
> One of the priorities in this task is to add a read option for the following fields:
> 
>     * `vcd_org_vdc.compute_capacity.used` to reflect how much of the VDC allocation is used currently
> 
>     * `vcd_org_vdc.compute_capacity.reserved`

These fields added.

**Comment by buec011** [comment]
Still missing for vcd_vapp_vm. We had an error today during an terraform apply for a new vcd_vapp_vm.
After we re-run the terraform apply, we got an error that the VM already exists instead of an update of the already build VM.

**Comment by vbauzysvmware** [comment]
Hi @buec011 we almost ready to publish PR for VM. So if you will be interested you can test it `fresh` :)

**Comment by buec011** [comment]
> Hi @buec011 we almost ready to publish PR for VM. So if you will be interested you can test it `fresh` :)

Nice to hear, I will do if my schedule allows it.

**Comment by vbauzysvmware** [comment]
https://github.com/terraform-providers/terraform-provider-vcd/pull/386

**Comment by buec011** [comment]
Tested #386: (Is it better to write my results here or at the PR)?

I added a wrong memory value (e.g. "1") such that I produced an API error:

```
vcd_vapp_vm.instance: Still creating... [1m30s elapsed]

Error: task did not complete successfully:  [400:BAD_REQUEST] - [ e981be7a-b463-47d3-a5aa-6fb0e7ca493f ] The virtual machine memory size must be a multiple of 4 MB.

  on main.tf line 45, in resource "vcd_vapp_vm" "instance":
  45: resource "vcd_vapp_vm" "instance" {
```

I changed the memory to "1024" and rerun the terraform apply.
I found multiple issues:
a) It wants to rebuild the VM? Is this really needed. I expected an in place update of the VM?
b) metadata and guest_properties should not be marked as removed. Same or "known after apply"
c) ip and mac in network block shloud not change.

```
> terraform apply
vcd_vapp.vapp: Refreshing state... [id=urn:vcloud:vapp:decb29d1-3463-47b3-a6e6-a304652b41a1]
vcd_vapp_vm.instance: Refreshing state... [id=urn:vcloud:vm:ebbb7549-20af-4a31-b94a-3d682d347510]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # vcd_vapp_vm.instance is tainted, so must be replaced
-/+ resource "vcd_vapp_vm" "instance" {
        accept_all_eulas               = true
        catalog_name                   = "asys-linux"
        computer_name                  = "buec011-test1"
        cpu_cores                      = 1
        cpus                           = 1
      + description                    = (known after apply)
        expose_hardware_virtualization = false
      - guest_properties               = {} -> null
      ~ href                           = "https://access.devcloud.arvato-systems.de/api/vApp/vm-ebbb7549-20af-4a31-b94a-3d682d347510" -> (known after apply)
      ~ id                             = "urn:vcloud:vm:ebbb7549-20af-4a31-b94a-3d682d347510" -> (known after apply)
      + ip                             = (known after apply)
      + mac                            = (known after apply)
      ~ memory                         = 2048 -> 1024
      - metadata                       = {} -> null
        name                           = "buec011-test1"
        power_on                       = true
      ~ storage_profile                = "Intermediate" -> (known after apply)
        template_name                  = "asys_centos_7_latest"
        vapp_name                      = "buec011-test1"

      ~ network {
          ~ ip                 = "10.65.67.100" -> (known after apply)
            ip_allocation_mode = "POOL"
            is_primary         = true
          ~ mac                = "00:50:56:0b:01:41" -> (known after apply)
            name               = "test-network-01"
            type               = "org"
        }
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```

HCL used:

```
variable "hostname" {
  description = "Hostname for the instance"
  default     = "please_fill_me"
}

variable "cpu" {
  description = "Number of CPUs for the instance"
  default     = "please_fill_me"
}

variable "core" {
  description = "Number of cores per CPU for the instance"
  default     = "please_fill_me"
}

variable "memory" {
  description = "Amount of memory for the instance"
  default     = "please_fill_me"
}

variable "network" {
  description = "Name of the network for the instance"
  default     = "please_fill_me"
}

variable "private_ip" {
  description = "Private IP for the instance. If nothing set take automatically generated IP."
  default     = null
}

variable "image_catalog" {
  description = "Name of the catalog of the image/template"
  default     = "please_fill_me"
}

variable "image_string" {
  description = "Name of the image/template"
  default     = "please_fill_me"
}

resource "vcd_vapp" "vapp" {
  name = var.hostname
}

resource "vcd_vapp_vm" "instance" {
  name          = var.hostname
  computer_name = var.hostname
  vapp_name     = vcd_vapp.vapp.name
  catalog_name  = var.image_catalog
  template_name = var.image_string
  cpus          = var.cpu
  cpu_cores     = var.core
  memory        = var.memory

  network {
    type               = "org"
    name               = var.network
    ip_allocation_mode = var.private_ip == null ? "POOL" : "MANUAL"
    ip                 = var.private_ip
    is_primary         = true
  }
}

output "instance_id" {
  value = vcd_vapp_vm.instance.id
}

output "no_nat_private_ip" {
  value = vcd_vapp_vm.instance.network[0].ip
}
```

**Comment by buec011** [comment]
Related, because vcd_vapp_vm has the same issues: https://github.com/terraform-providers/terraform-provider-vcd/issues/387

**Comment by dataclouder** [comment]
@buec011  Thanks for your feedback.
Your problem is the result of an error during apply, which automatically taints the resource.
You can solve the issue by running
```
 terraform untaint vcd_vapp_vm.instance
```
and then running `plan` and `apply` once more after correcting the value for memory.
You will see that the resource will not be re-created, but updated in place.

To mitigate the risk of such errors, I am adding a validation for the value of `memory`. It will fail immediately if the value is not a multiple of 1024, which would save some time.

**Comment by buec011** [comment]
Hi @dataclouder 
> Your problem is the result of an error during apply, which automatically taints the resource.

Is this a terraform or a provider feature? Asking because I did not noticed during use of other provider.

**Comment by dataclouder** [comment]
I don't know for sure, but I assume tainting is Terraform expected behavior after a failure.

**Comment by lvirbalas** [comment]
@buec011 I'd like to close this issue as reads have been implemented for most resourses in 2.5.0 and, with recent merge of #386, for VM as well (in master). Would you file separate issue(s) for the other topics (like in-place updates)?

**Comment by buec011** [comment]
>  I'd like to close this issue as reads have been implemented for most resourses in 2.5.0 and, with recent merge of #386, for VM as well (in master)

OK. I will open new issues for further improvments and bugs
